### PR TITLE
[TFLite] Add TABLE operator for LUT-based operators lowering

### PR DESCRIPTION
**Pull request by Tessil**
Hi,

This PR adds a TFLite TABLE operator which is similar to the TABLE operator of the [TOSA specification](https://developer.mlplatform.org/w/tosa/). The operator takes int8 or int16 inputs and look them up into the table associated to the operator to produce the outputs.

This operator can be used to lower non-linear quantized functions like the exponential to a LUT in the range of the quantized range. The following [proof-of-concept](https://github.com/Tessil/tensorflow/commit/84eedece6d8d409a040362a375d669a3a0997abb) commit quantizes the EXP operator by generating a LUT in the input range of the operator and replace the it by a TABLE operator in the exported TFLite model.

This PR only add the operator and don't provide any transformation yet, these will be part of a different PR. The `gen_lut` function also has been extended to support int8->int8, int8->int16 and int16->int8 tables in addition of the previously supported int16->int16 table.

Thibaut

### CODE REVIEW

**Comment by abattery** [comment]
Thanks for your contribution.

The new table operator will not be used for a part of the TF to TFL op lowering so it is hard to be used for the actual use cases. Are there any future plans for that? If now, how about just having this table operator as a custom op?

**Comment by gbaned** [comment]
@Tessil  Can you please check @abattery's comments and keep us posted ? Thanks!

**Comment by abattery** [comment]
Since this operator is somewhat in an experimental stage, how about adding this operator as a custom op as a first step? Until this custom op will resolve the unsupported cases or unlock the new opportunities with concrete examples since the builtin op addition will demand most of the android developers for the extra binary size requirement.

**Comment by Tessil** [comment]
Sorry for my late answer. I think it would be easier for now to leave the PR on the side as there are still some discussions in progress regarding the advantages of a separate TABLE operator compared to generating the LUT inside the `Prepare` method of each operator. Once we have a clearer understanding on how to move on with all this, we could eventually add the operator as custom op as first step.

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@Tessil  Any update on this PR? Please. Thanks!

**Comment by Tessil** [comment]
Hi @gbaned ,

The PR is put in suspend for now as some discussions are still required internally and with some of the TFLite team members to check if we move forward with a separate TABLE operator or not. Sorry for that.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

File: tensorflow/lite/kernels/internal/common.h [hunk 1]
```diff
@@ -279,81 +281,126 @@ inline Integer FloorLog2(Integer n) {
...
-  table[num - 1] = std::min<float>(
-      std::max<float>(TfLiteRound(func(max) * 32768.0f), -32768.0f), 32767.0f);
 }
 
-// int16_t func table lookup, e.g., lookup exp() and 1/(1+x) used in softmax
-inline int16_t generic_int16_table_lookup(int16_t value, const int16_t* lut) {
-  // 512 base value, lut[513] only for calculate slope
-  uint16_t index = static_cast<uint16_t>(256 + (value >> 7));
+// LUT must have 513 values
+template <typename LutOutT>
```
Comment (jianlijianli) [inline tensorflow/lite/kernels/internal/common.h:359]: Nit: maybe LutOutT should just be LutT? Usually when we see a "output" type, we expect something on the "input".

File: tensorflow/lite/kernels/internal/common.h [hunk 2]
```diff
@@ -279,81 +281,126 @@ inline Integer FloorLog2(Integer n) {
...
-                                32767.0);
-  }
-  table[num - 1] = std::min<double>(
-      std::max<double>(TfLiteRound(func(max) * 32768.0), -32768.0), 32767.0);
+// The size of the LUT depends on the type of input. For int8 inputs a simple
+// 256 entries LUT is used. For int16 inputs the high 9 bits are used for
+// indexing and the 7 remaining bits are used for interpolation. We thus use a
+// 513-entries LUT for int16 cases, 512 for the 9-bit indexing and 1 extra entry
+// to interpolate the last value.
+template <typename LutInT>
```
Comment (jianlijianli) [inline tensorflow/lite/kernels/internal/common.h:289]: Nit: can we rename the LutInT to just LutT? I feel the concept of input and output for a LUT is a bit blury.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td [hunk 3]
```diff
@@ -1962,6 +1962,25 @@ def TFL_LogOp: TFL_Op<"log", [
   let hasFolder = 1;
 }
 
+def TFL_TableOp: TFL_Op<"table", [
```
Comment (jianlijianli) [inline tensorflow/compiler/mlir/lite/ir/tfl_ops.td:1965]: Is this file needed? It's not very clear what's the end-to-end workflow with a dedicated Table operator.

**Review by jianlijianli — COMMENTED** [review]
Thanks Tessil, please take a look at the comments.

File: tensorflow/lite/kernels/internal/common.h [hunk 4]
```diff
@@ -279,81 +281,126 @@ inline Integer FloorLog2(Integer n) {
...
+  // Round and convert from Q0.(x + 7) to Q0.x
+  const int delta = (slope * offset + 64) >> 7;
 
   // Q0.15 + Q0.15
-  return base + delta;
+  return static_cast<LutOutT>(base + delta);
+}
+
+// int16_t -> int16_t table lookup with interpolation
+// LUT must have 513 values
```
[the remark this answers is not in the document]
Comment (Tessil) [reply tensorflow/lite/kernels/internal/common.h:383]: The structures in [types.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/types.h) use simple pointers for tables, we can't deduce the size from these . We could either use something similar to to `std::span`/`absl::Span` but the first requires C++20 and the second Abseil (which from my understanding isn't part of the TFLite dependencies).

Note that `generic_int16_table_lookup` didn't check either (neither does any other accesses on pointer types in [types.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/types.h)) but we can eventually check to improve that.

File: tensorflow/lite/kernels/table_test.cc [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                             float output_max) {
+  static_assert(
+      std::is_same<T, int8_t>::value || std::is_same<T, int16_t>::value,
+      "T must be an int8_t or int16_t.");
+
+  const float range_sum = (input_max - input_min) + (output_max - output_min);
+  if (std::is_same<T, int8_t>::value) {
+    return range_sum / 256.0f;
+  } else {
+    return range_sum / 512.0f;
```
[the remark this answers is not in the document]
Comment (Tessil) [reply tensorflow/lite/kernels/table_test.cc:74]: I extended the comment to explain the reason.

File: tensorflow/lite/kernels/internal/common.h (hunk 1 above)
Comment (Tessil) [reply tensorflow/lite/kernels/internal/common.h:359]: The `LutOutT` in `lut_lookup_with_interpolation` and `LutInT` in `lut_size` are related to the ones used in `gen_lut`. I used these instead of just `LutT` to avoid any confusion on which one should be used for these functions.

I can eventually extend the comment in `gen_lut` to better explain the `LutInT`/`LutOutT` concept.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td (hunk 3 above)
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.td:1965]: Yes, we need to be able to create a `mlir::TFL::TableOp` in the the converter. You can check the [proof-of-concept](https://github.com/Tessil/tensorflow/commit/84eedece6d8d409a040362a375d669a3a0997abb#diff-9f4911916593feb3005c005582353a39430b45e161458a36673ea1bfac78b26fR106) I linked above that demonstrates the expected end-to-end workflow. We convert a `mlir::TFL::ExpOp` with a quantized input/output range to a `mlir::TFL::TableOp`.

File: tensorflow/lite/kernels/internal/common.h (hunk 2 above)
Comment (Tessil) [reply tensorflow/lite/kernels/internal/common.h:289]: See comment on `lut_lookup_with_interpolation`.

**Comment by abattery** [comment]
There is no enough discussion on this new approach's additions. How about closing this PR until we have enough concensus on this?

File: tensorflow/lite/kernels/internal/common.h (hunk 2 above)
Comment (jianlijianli) [reply tensorflow/lite/kernels/internal/common.h:289]: Thanks.

File: tensorflow/lite/kernels/internal/common.h (hunk 1 above)
Comment (jianlijianli) [reply tensorflow/lite/kernels/internal/common.h:359]: Thanks. Make sense.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td (hunk 3 above)
Comment (jianlijianli) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.td:1965]: Thanks for the end-to-end example. A high level question: LUT is replacing reference and optimized implementation?

**Comment by jianlijianli** [comment]
Thanks Tessil, please take a look at the comments.

Thanks abattery, this is on on-going discussion so let's not close yet :)

File: tensorflow/lite/kernels/internal/common.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Tessil) [reply tensorflow/lite/kernels/internal/common.h:383]: We check the size of the table tensor in Table::Prepare. We can eventually double-check again in Table::Eval though it may be a bit overzealous.

File: tensorflow/compiler/mlir/lite/ir/tfl_ops.td (hunk 3 above)
Comment (Tessil) [reply tensorflow/compiler/mlir/lite/ir/tfl_ops.td:1965]: Yes as there is no other way currently. The choice between reference or optimized kernels is decided at runtime when creating the `tf.lite.Interpreter` (a same *.tflite file can be run with the reference or optimized kernels without problem) . On the other hand, the `TABLE` operator is created during the generation of the *.tflite file we thus can't have a different implementation for the initial converted operator once it's done.

**Comment by jianlijianli** [comment]
Thanks Tessil. Sorry I missed your reply (didn't realized that github folds your reply right after each comment and I was only checking there were no new discussions below https://github.com/tensorflow/tensorflow/pull/45342#issuecomment-845282707)

Thanks for fixing all the issues and explaining the bigger picture.

**Review by jianlijianli — APPROVED** [review]

**Comment by Tessil** [comment]
Thank you very much for the review @jianlijianli. I fixed the merge conflicts that were remaining, the PR will need re-approval.

**Comment by abattery** [comment]
Adding @karimnosseir @miaout17 in the thread for visibility of the external schema changes.

**Comment by jianlijianli** [comment]
Thanks @abattery for flagging potential challenges and suggesting custom op approach.

Tessil, following your comment in https://github.com/tensorflow/tensorflow/pull/45342#issuecomment-749597630, and since we already looked into the technical details, is it a good time to move this to a custom op? Once the PR is in and well tested, the promotion should be simple.

@karimnosseir @miaout17 , please take a look as well.

**Comment by Tessil** [comment]
I have to look a bit on how to adapt the [proof-of-concept](https://github.com/Tessil/tensorflow/commit/84eedece6d8d409a040362a375d669a3a0997abb) as if we move the TABLE operator to a custom op we can't have a `TFL_TableOp` in `tfl_ops.td` and thus a `mlir::TFL::TableOp`.

I'll try to adapt the proof-of-concept to use `mlir::TFL::CustomOp` and see how it works.

**Comment by Tessil** [comment]
@jianlijianli I updated the PR and moved the TABLE operator from a built-in op to a custom op.

I also updated the [proof-of-concept](https://github.com/Tessil/tensorflow/commit/b0363b833e510d8fa4596f0007acbad888496ea1) to illustrate the lowering using a `mlir::TFL::CustomOp` instead of a `mlir::TFL::TableOp`.

**Comment by gbaned** [comment]
@Tessil  Can you please resolve conflicts? Thanks!

**Comment by Tessil** [comment]
@gbaned Thanks I fixed the conflicts..

Note that with the removal of TensorFlow Lite Micro from this repository, the change I did in [tensorflow/lite/micro/kernels/softmax.cc](https://github.com/tensorflow/tensorflow/pull/45342/commits/3cde0d4288b355443483efd5cb19eaaf23b6c510#diff-91be54c4ee3217090907f1dbfe935d90945b5a253615ae542e3700305e115ff8) will need to be done in https://github.com/tensorflow/tflite-micro when syncing the TF repository after the change has been merged. Not sure what is the best way to proceed with this.

**Comment by gbaned** [comment]
@jianlijianli  Can you please assist on above comments from @Tessil. Thanks!

**Comment by jianlijianli** [comment]
Thanks Tessil for making the change.

> the change I did in tensorflow/lite/micro/kernels/softmax.cc will ....
If I understand correctly, the change in softmax is only refactoring (so the common functions can support the new TABLE operator). Are there any tangible change in the implementation?

**Comment by Tessil** [comment]
Yes, the change is only a refactoring to adapt to the new more flexible `gen_lut` function signature. There is no tangible change in the implementation itself, the results will be exactly the same for the given parameters.

**Review by jianlijianli — APPROVED** [review]
Thanks Tessil.

**Comment by Tessil** [comment]
@jianlijianli I fixed an implicit cast warning which caused an error in one of the CI. The PR will need re-approval, thanks!

**Review by jianlijianli — APPROVED** [review]
Thanks Tessil. Gtihub still shows build failure on Intel oneDNN but it appear to be unrelated.

**Comment by abattery** [comment]
Is it possible to remove the new custom op from the default op registration? Instead, we can keep it as a part of `"custom_ops` target under the tensorflow/lite/BUILD.

File: tensorflow/lite/kernels/BUILD [hunk 6]
```diff
@@ -637,6 +637,7 @@ BUILTIN_KERNEL_SRCS = [
     "strided_slice.cc",
     "sub.cc",
     "svdf.cc",
+    "table.cc",
```
Comment (abattery) [inline tensorflow/lite/kernels/BUILD:640]: Is it possible to remove the new custom op from the default op registration? Instead, we can keep it as a part of custom_ops target under the tensorflow/lite/BUILD file.

Most of common TFLite users do not requires a new table custom op for now and we would like to keep the TFLite library as compact as possible.

**Comment by Tessil** [comment]
I can remove it from the default registration and add it to the [custom_ops](https://github.com/tensorflow/tensorflow/blob/81908ee16f7e0efa611c408553271e48f68c9041/tensorflow/lite/kernels/BUILD#L779) library but out of curiosity are there specific builds generated with these custom ops? And how should we compile such builds (as the `custom_ops` library target doesn't seem to be referenced anywhere as dependency outside of the tests)?
